### ISSUE TITLE
regression: Postgres source never transitions to a "stalled" state after error, replication has stopped 

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#17132]
### What version of Materialize are you using?

 v0.40.0-dev (70b6ba169)

### How did you install Materialize?

Docker image

### What is the issue?

Nightly has a job, "Checks Parallel + DROP/CRETE replica" that has never failed before. What this job does is:

1. Runs all the `initialize()` and the first `manipulate()` .td fragments from `misc/python/materialize/checks` in parallel (that is, using concurrent testdrive invocations) against a Mz instance.
2. Once they are all complete, runs:
```
           DROP CLUSTER REPLICA default.r1;
           CREATE CLUSTER REPLICA default.r1 SIZE '1';
```
3. Runs the second `manipulate()` and the `validate()` .td fragments , again in parallel.

This fails in Nightly and on a ci scratch machine as follows:

```
2:1: error: non-matching rows: expected:
--
  | [["stalled", "true"], ["stalled", "true"]]
  | got:
  | [["running", "<null>"], ["stalled", "true"]]
  | Poor diff:
  | + running <null>
  | - stalled true
  |  
  | \|
  | 1 \|
  | 2 \| > SELECT status, error ~* 'publication .+ does not exist'
  | \| ^
```

In other words, one of the Checks creates two postgres sources and then issues `DROP PUBLICATION` on the Postgres side to make both sources error out. However, only one of the two sources errors out:

```
materialize=> SELECT name,  status, error, error ~* 'publication .+ does not exist'
                  FROM mz_internal.mz_source_statuses
                  WHERE name LIKE 'source_errors_source%';
         name          | status  |                                  error                                   | ?column? 
-----------------------+---------+--------------------------------------------------------------------------+----------
 source_errors_sourcea | running |                                                                          | 
 source_errors_sourceb | stalled | db error: ERROR: publication "source_errors_publicationb" does not exist | t
```

The problem is that `source_errors_sourcea` is listed as running, but in fact is equally borked as its publication has also been dropped by the test on the Postgres side. The source is empty:

```
materialize=> select * from source_errors_sourcea;
--
(0 rows)
```

Even though the postgres-side table contains rows. Inserting a new rows on the postgres side is also not reflected on the Mz side, the source remains empty.

The expectation is that both sources will be "stalled". There are no OOMs or panics during this test. The log also does not report any errors relating to that source, only to the other one that errored correctly.

To reproduce:
1. Start a scratch machine of type `ci`.
2. Run:
```
cd test/platform-checks
while bash -c -e "./mzcompose down -v ; ./mzcompose run default --scenario=NoRestartNoUpgrade --execution-mode=parallel" ;  do : ; done
```
Until it fails.

@bkirwi FYI

**Comment by philip-stoev** [comment]
Seen for the first time Dec 13th https://buildkite.com/materialize/nightlies/builds/1485#01850b79-6475-4c68-a16d-095a4913e402

**Comment by bkirwi** [comment]
It looks like the Dec 13th failure was a different job, checks parallel + restart redpanda & debezium. So this is appears to be an issue evinced by the specific postgres check and not the other disruptions that are happening at the same time.

I'm not certain how to attack this one. It sounds like we get a failure in one of every hundred test executions or so, assuming this issue has been there since the beginning. It's possible we could make it more likely with a very focussed test.

We should definitely check in on this once `halt!` has been removed from the source pipeline. (If it's not a process lifecycle issue, that implies we have a pg protocol handling bug and not a error reporting bug.)

**Comment by aljoscha** [comment]
I believe I found the problem! This is what I get when reproducing the problem:

```
materialize=> SELECT *, error ~* 'publication .+ does not exist'
                  FROM mz_internal.mz_source_statuses
                  WHERE name LIKE 'source_errors_source%';
  id  |         name          |   type   |   last_status_change_at    | status  |                                  error                                   | details | ?column?
------+-----------------------+----------+----------------------------+---------+--------------------------------------------------------------------------+---------+----------
 u101 | source_errors_sourcea | postgres | 2023-01-19 13:25:54.552+00 | running |                                                                          |         |
 u118 | source_errors_sourceb | postgres | 2023-01-19 13:27:51.6+00   | stalled | db error: ERROR: publication "source_errors_publicationb" does not exist |         | t
(2 rows)

materialize=> show sources;
         name          |   type    | size
-----------------------+-----------+------
 source_errors_sourcea | postgres  | 4
 source_errors_sourceb | postgres  | 4
 source_errors_tablea  | subsource |
 source_errors_tableb  | subsource |
(4 rows)

materialize=> select * from source_errors_tableb;
ERROR:  Source error: u118: db error: ERROR: publication "source_errors_publicationb" does not exist
materialize=> select * from source_errors_tablea;
 f1
----
 1
 2
 3
 4
(4 rows)

materialize=>
```

It seems that `sourcea` is in fact healthy, and can be queried. The crucial thing is that for postgres sources the toplevel source itself doesn't have any data, it's only the tables that are read by them that have data. The subsources.

It seems somehow the dropping of the publication is not going through. I can still manually go into postgres and insert a row into the table. And it _will_ show up in our source table.

**Comment by aljoscha** [comment]
@philip-stoev ☝️ 

If you look at the machine when the error reproduces, you can see that the CPUs are fully pegged. Could be that postgres is soiling itself in that situation, somehow.

But it seems this is not a bug in our code/I don't know what we could do to fix the tests. How should we proceed?

**Comment by aljoscha** [comment]
gentle ping

**Comment by philip-stoev** [comment]
I made a comment to this but it has been lost apparently. In my case the source is both empty, not errored out and future inserts are not reflected in it. So I think it is a genuine bug in the product.

More specifically, as in the original comment, those two queries return results like this:

```
materialize=> SELECT name,  status, error, error ~* 'publication .+ does not exist'
                  FROM mz_internal.mz_source_statuses
                  WHERE name LIKE 'source_errors_source%';
         name          | status  |                                  error                                   | ?column? 
-----------------------+---------+--------------------------------------------------------------------------+----------
 source_errors_sourcea | running |                                                                          | 
 ...
```
and
```
materialize=> select * from source_errors_sourcea;
--
(0 rows)
```

Those two things can not be like that in a consistent situation -- either the source has errored out, or it should have some data. You can not have a source that is both empty and not errored out.

**Comment by aljoscha** [comment]
I mentioned this in my message above, but did you check `source_errors_tablea`, NOT `source_errors_sourcea`. In your example query you use the latter, which will always be empty.

**Comment by philip-stoev** [comment]
I am taking this upon myself to reproduce once more and see if the test needs fixing . Thank you and apologies for the back and forth.

**Comment by aljoscha** [comment]
Oh not at all! All this is very helpful, I just want to weed out issues that are no longer a problem (at least in the storage code itself). 😅

**Comment by mjibson** [comment]
I'm not sure if this should be a new issue or is this one. I have recently seen many test flakes of "Checks + restart Redpanda & Debezium" ([example](https://buildkite.com/materialize/tests/builds/50081#018660fb-e050-4c2a-9a79-fda4f1555ccd)) and would like to nudge another look at this.

**Comment by aljoscha** [comment]
@mjibson That issue is https://github.com/MaterializeInc/materialize/issues/17543, which has been fixed (hopefully!) by https://github.com/MaterializeInc/materialize/pull/17786.

I checked, and your CI run is from before the fix, but please do report if you still see this. The symptom for this fixed bug is when you see a `cannot serve requested as_of` in the logs (`services.log` under the Artifacts tab).

**Comment by philip-stoev** [comment]
This has not been seen in a while so I am closing it until it happens again in the CI.
